File: src/main/scala/io/izzel/taboolib/module/packet/channel/ChannelExecutor.java
package io.izzel.taboolib.module.packet.channel;

import io.izzel.taboolib.module.packet.Packet;
import io.izzel.taboolib.module.packet.TPacketHandler;
import io.netty.channel.Channel;
import io.netty.channel.ChannelDuplexHandler;
import io.netty.channel.ChannelHandlerContext;
import io.netty.channel.ChannelPromise;
import org.bukkit.entity.Player;

import java.util.Collection;
import java.util.ConcurrentModificationException;
import java.util.concurrent.ExecutorService;
import java.util.concurrent.Executors;

/**
 * @Author 坏黑
 * @Since 2018-10-28 14:34
 */
public abstract class ChannelExecutor {

    private final ExecutorService addChannelService = Executors.newSingleThreadExecutor();
    private final ExecutorService removeChannelService = Executors.newSingleThreadExecutor();

    public abstract void sendPacket(Player player, Object packet);

    public abstract Channel getPlayerChannel(Player player);

    public void addPlayerChannel(Player player) {
        addChannelService.submit(() -> {
            getPlayerChannel(player).pipeline().addBefore("packet_handler", "taboolib5_packet_handler", new ChannelHandler(player));
        });
    }

    public void removePlayerChannel(Player player) {
        removeChannelService.submit(() -> {
            Channel playerChannel = getPlayerChannel(player);
            if (playerChannel.pipeline().get("taboolib5_packet_handler") != null) {
                playerChannel.pipeline().remove("taboolib5_packet_handler");
            }
        });
    }

    static class ChannelHandler extends ChannelDuplexHandler {

        private final Player player;

        public ChannelHandler(Player player) {
            this.player = player;
        }

        @Override
        public void write(ChannelHandlerContext channelHandlerContext, Object o, ChannelPromise channelPromise) throws Exception {
            try {
                if (TPacketHandler.getListeners().stream().flatMap(Collection::stream).anyMatch(packetListener -> !packetListener.onSend(player, o) || !packetListener.onSend(player, new Packet(o)))) {
                    return;
                }
            } catch (ConcurrentModificationException ignore) {
            } catch (Exception e) {
                e.printStackTrace();
            }
            super.write(channelHandlerContext, o, channelPromise);
        }

        @Override
        public void channelRead(ChannelHandlerContext channelHandlerContext, Object o) throws Exception {
            try {
                if (TPacketHandler.getListeners().stream().flatMap(Collection::stream).anyMatch(packetListener -> !packetListener.onReceive(player, o) || !packetListener.onReceive(player, new Packet(o)))) {
                    return;
                }
            } catch (ConcurrentModificationException ignore) {
            } catch (Exception e) {
                e.printStackTrace();
            }
            super.channelRead(channelHandlerContext, o);
        }
    }
}

File: module-nms/src/main/kotlin/taboolib/module/nms/ChannelExecutor.kt
package taboolib.module.nms

import io.netty.channel.Channel
import org.bukkit.entity.Player
import org.bukkit.event.player.PlayerJoinEvent
import org.bukkit.event.player.PlayerQuitEvent
import taboolib.common.platform.Platform
import taboolib.common.platform.PlatformSide
import taboolib.common.platform.SubscribeEvent
import taboolib.common.platform.info
import taboolib.common.reflect.Reflex.Companion.reflex
import taboolib.common.reflect.Reflex.Companion.reflexInvoke
import java.util.concurrent.Executors

/**
 * @author 坏黑
 * @since 2018-10-28 14:34
 */
@PlatformSide([Platform.BUKKIT])
object ChannelExecutor {

    private val addChannelService = Executors.newSingleThreadExecutor()
    private val removeChannelService = Executors.newSingleThreadExecutor()

    fun getPlayerChannel(player: Player): Channel {
        return if (MinecraftVersion.isUniversal) {
            player.reflex<Channel>("entity/connection/connection/channel")!!
        } else {
            player.reflex<Channel>("entity/playerConnection/networkManager/channel")!!
        }
    }

    fun addPlayerChannel(player: Player) {
        addChannelService.submit {
            try {
                getPlayerChannel(player).pipeline().addBefore("packet_handler", "taboolib6_packet_handler", ChannelHandler(player))
            } catch (ex: Throwable) {
                ex.printStackTrace()
            }
        }
    }

    fun removePlayerChannel(player: Player) {
        removeChannelService.submit {
            try {
                val playerChannel = getPlayerChannel(player)
                if (playerChannel.pipeline()["taboolib6_packet_handler"] != null) {
                    playerChannel.pipeline().remove("taboolib6_packet_handler")
                }
            } catch (ex: Throwable) {
                ex.printStackTrace()
            }
        }
    }

    @SubscribeEvent
    internal fun e(e: PlayerJoinEvent) {
        addPlayerChannel(e.player)
    }

    @SubscribeEvent
    internal fun e(e: PlayerQuitEvent) {
        removePlayerChannel(e.player)
    }
}